src/expr/benches/expr.rs

```diff
@@ -0,0 +1,158 @@
+    let i1 = || InputRefExpression::new(DataType::Int32, 1).boxed();
+    c.bench_function("expr/inputref", |bencher| {
+        let inputref = i0();
+        bencher.iter(|| inputref.eval(&input).unwrap())
+    });
+    c.bench_function("expr/constant", |bencher| {
+        let constant = LiteralExpression::new(DataType::Int32, Some(1_i32.into()));
+        bencher.iter(|| constant.eval(&input).unwrap())
+    });
+    c.bench_function("expr/add/i32", |bencher| {
+        let add = new_binary_expr(Type::Add, DataType::Int32, i0(), i1()).unwrap();
+        bencher.iter(|| add.eval(&input).unwrap())
+    });
+    c.bench_function("expr/mul/i32", |bencher| {
+        let mul = new_binary_expr(Type::Multiply, DataType::Int32, i0(), i1()).unwrap();
+        bencher.iter(|| mul.eval(&input).unwrap())
+    });
+    c.bench_function("expr/eq/i32", |bencher| {
+        let mul = new_binary_expr(Type::Equal, DataType::Int32, i0(), i1()).unwrap();
+        bencher.iter(|| mul.eval(&input).unwrap())
+    });
+    c.bench_function("expr/sum/i32", |bencher| {
+        let mut sum =
+            create_agg_state_unary(DataType::Int32, 0, AggKind::Sum, DataType::Int64, false)
+                .unwrap();
+        bencher.iter(|| sum.update_multi(&input, 0, 1024).unwrap())
+    });
+}
+
+/// Evaluate on raw Rust array.
+///
+/// This could be used as a baseline to compare and tune our expressions.
+fn bench_raw(c: &mut Criterion) {
+    // ~55ns
+    c.bench_function("raw/sum/i32", |bencher| {
+        let a = (0..CHUNK_SIZE as i32).collect::<Vec<_>>();
+        bencher.iter(|| a.iter().sum::<i32>())
+    });
+    // ~90ns
+    c.bench_function("raw/add/i32", |bencher| {
+        let a = (0..CHUNK_SIZE as i32).collect::<Vec<_>>();
+        let b = (0..CHUNK_SIZE as i32).collect::<Vec<_>>();
+        bencher.iter(|| {
+            a.iter()
+                .zip(b.iter())
+                .map(|(a, b)| a + b)
+                .collect::<Vec<_>>()
+        })
+    });
+    // ~600ns
+    c.bench_function("raw/add/i32/zip_eq", |bencher| {
+        let a = (0..CHUNK_SIZE as i32).collect::<Vec<_>>();
+        let b = (0..CHUNK_SIZE as i32).collect::<Vec<_>>();
+        bencher.iter(|| {
+            itertools::Itertools::zip_eq(a.iter(), b.iter())
+                .map(|(a, b)| a + b)
+                .collect::<Vec<_>>()
+        })
+    });
+    // ~950ns
+    c.bench_function("raw/add/Option<i32>/zip_eq", |bencher| {
+        let a = (0..CHUNK_SIZE as i32).map(Some).collect::<Vec<_>>();
+        let b = (0..CHUNK_SIZE as i32).map(Some).collect::<Vec<_>>();
+        bencher.iter(|| {
+            itertools::Itertools::zip_eq(a.iter(), b.iter())
+                .map(|(a, b)| match (a, b) {
+                    (Some(a), Some(b)) => Some(a + b),
+                    _ => None,
+                })
+                .collect::<Vec<_>>()
+        })
+    });
+    // ~2100ns
+    c.bench_function("raw/add/Option<i32>/zip_eq,checked", |bencher| {
```

Comment:
```suggestion
    c.bench_function("raw/add/Option<i32>/zip,checked", |bencher| {
```

Reply:
This is intended. `zip_eq` introduced a check for each yield which greatly hurts the performance. We should switch back to `zip` and check the length at the beginning. 🤪

Reply:
Sorry my fault. It should use. 🥵